File: 防火墙项目/firewallclient-nodejs/classes/configCNCRules.js
var sendInfo = require('../tools/sendInfo.js');
var dbOperations = require('../tools/dbOperations.js');
var configInfo = require('../tools/listenConfigInfo.js');

/**
 * 配置连接数控制规则
 * @param {*设备IP地址string} devIp 
 * @param {*连接数string} conlimit 
 * @param {*源IP地址string(any表示任意)} srcIp 
 * @param {*目的IP地址string} dstIp 
 * @param {*源端口string} sport 
 * @param {*目的端口string} dport 
 * @param {*true为添加，false为删除bool} addOrDelete 
 */

function configCNCRules(devIp, conlimit, srcIp, dstIp, sport, dport, addOrDelete) {
    if (devIp == "0.0.0.0")
        return false;
    var rule, CNCRules, sqlRule;
    if (srcIp != "any") {
        rule = "iptables -A FORWARD -p tcp --syn -s " + srcIp + " --sport " + sport + " -d " + dstIp + " --dport " + dport +
            " -m connlimit --connlimit-above " + conlimit;
    }
    if (srcIp == "any") {
        rule = "iptables -A FORWARD -p tcp --syn --sport " + sport + " -d " + dstIp + " --dport " + dport +
            " -m connlimit --connlimit-above " + conlimit;
    }

    var CNCLogRule = rule + " -j LOG";

    if (addOrDelete) {
        CNCRules = "CNC1" + CNCLogRule + " && " + rule;
        sqlRule = "INSERT INTO cnc VALUES " + "('" + devIp + "','" + conlimit + "','" + srcIp + "','" + dstIp + "','" + sport + "','" + dport + "')";
        dbOperations.dbOperation(sqlRule);
        sendInfo.sendConfigInfo(devIp, CNCRules);
        return (configInfo.listenConfigInfo());
    }
    if (!addOrDelete) {
        CNCRules = "CNC0" + CNCLogRule + " && " + rule;
        sqlRule = "DELETE FROM cnc WHERE (devIP='" + devIp + "' and connlimit=" + conlimit + " and srcIP='" + srcIp + "' and dstIP='" + dstIp + "' and sport='" + sport + "' and dport='" + dport + "')";
        dbOperations.dbOperation(sqlRule);
        sendInfo.sendConfigInfo(devIp, CNCRules);
        return (configInfo.listenConfigInfo());
    }
}
exports.configCNCRules = configCNCRules;

File: QT-firewall/firewall-test/configcncrules.cpp
#include "configcncrules.h"

ConfigCNCRules::ConfigCNCRules(const std::string &dev_IP, const std::string &fw_mac, QObject *parent=nullptr)
    :FwOrProIP(dev_IP),fwMAC(fw_mac),QCommInfo(parent)
{
    fileOP = new FileOperation();
    this->setBindInfo(getLocal_IP(),30333);
}

ConfigCNCRules::~ConfigCNCRules()
{
    if(fileOP != nullptr)
        delete fileOP;
    fileOP = nullptr;
}

static bool fileFlag = true;
static std::string file_rule("");

void ConfigCNCRules::ChangeConnectionControl(const std::string &srcIP, const std::string &dstIP, const std::string &sport, const std::string &dport, int number, bool addOrdel)
{
    std::string pro = "CNC ConnectionControl";
    std::string src = " -s "+srcIP;
    std::string dst = " -d "+dstIP;
    if(srcIP == "any")
        src.assign("");
    if(dstIP == "any")
        dst.assign("");
    std::string flag = "";std::string rule("");
    if(addOrdel)
    {
        flag.assign("CNC1");
        fileFlag = true;
    }
    else
    {
        flag.assign("CNC0");
        fileFlag = false;
    }

    rule = flag+"iptables -A FORWARD -p tcp --syn"+src+" --sport "+sport+dst+" --dport "
            +dport+" -m connlimit --connlimit-above "+std::to_string(number);
    file_rule = pro+" "+fwMAC + " "+srcIP+" "+dstIP + " "+sport+" "+dport+" "+std::to_string(number);
    this->sendCmd(FwOrProIP,22222,CommInfo::iptablesCmdTypeByte,rule);
}

void ConfigCNCRules::handResultData(const bool result, const char *commandTypeByte, const char *commandBytes, const int commandLength, const QString &srcIp, const int srcPort)
{
    if (result)
    {
        if (*commandTypeByte == *CommInfo::iptablesCmdTypeByte)
        {
            std::string data(commandBytes,commandLength);
            //std::cout << data << std::endl;
            if(data == "success")
            {
                if(fileFlag)
                    fileOP->addRule(file_rule);
                else
                    fileOP->deleteRule(file_rule);
                emit configResult(true);
            }
            if(data == "fail")
                emit configResult(false);
        }
    }
}
